### ISSUE TITLE
Denoise throws ValueError

### ISSUE THREAD

**Issue by AetherUnbound** [scikit-image/scikit-image#3449]
## Description

`_denoise_cy.pyx` throws a `ValueError` saying a `double` was expected but a `float` was received.

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/aether/programs/anaconda3/envs/cellprofiler/lib/python2.7/site-packages/skimage/restoration/_denoise.py", line 117, in denoise_bilateral
    bins, mode, cval)
  File "skimage/restoration/_denoise_cy.pyx", line 114, in skimage.restoration._denoise_cy._denoise_bilateral
ValueError: Buffer dtype mismatch, expected 'double' but got 'float'
```

## Way to reproduce

- [X] Code example
```
import numpy as np
import skimage.restoration

sigma = 16.0
sigma_range = .2
np.random.seed(0)
image = np.random.uniform(size=(100, 100)).astype(np.float32)
mask = np.ones(image.shape, bool)
mask[40:60, 45:65] = False
expected = skimage.restoration.denoise_bilateral(
    image=image,
    multichannel=False,
    sigma_color=sigma_range,
    sigma_spatial=sigma,
)
```
- [ ] Relevant images (if any)
- [X] Operating system and version: Ubuntu 16.04
- [X] Python version
```
>>> print(sys.version)
2.7.14 |Anaconda, Inc.| (default, Dec  7 2017, 17:05:42) 
[GCC 7.2.0]
```
- [X] scikit-image version (run `skimage.__version__`)
```
>>> skimage.__version__
'0.14.1'
```

**Comment by jni** [comment]
Holy moly! I believe #3110 introduced this: before, we would have upconverted your array to float64.

@hmaarrfk added some nice fused typedefs in #3253. We should fast-track that so that we can use Cython fused types for float everywhere doubles are used currently — including in `denoise_bilateral`.

**Comment by hmaarrfk** [comment]
@jni, I think you should just pull out the relevant file. I'll deal with the conflicts later.

**Comment by mlippie** [comment]
I just compiled scikit-image 0.14.3 from source, which was supposed to fix this issue. Unfortunately, the issue remains in CellProfiler testing (see https://github.com/CellProfiler/CellProfiler/pull/3661#issuecomment-430393640).

I ge this error log:
```
    def test_05_01_bilateral(self):
        '''test the smooth module with bilateral filtering'''
        sigma = 16.0
        sigma_range = .2
        np.random.seed(0)
        image = np.random.uniform(size=(100, 100)).astype(np.float32)
        mask = np.ones(image.shape, bool)
        mask[40:60, 45:65] = False
        expected = skimage.restoration.denoise_bilateral(
            image=image,
            multichannel=False,
            sigma_color=sigma_range,
>           sigma_spatial=sigma,
        )

tests/modules/test_smooth.py:232: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../../miniconda3/envs/cp-wxpython4-py2714/lib/python2.7/site-packages/skimage/restoration/_denoise.py:171: in denoise_bilateral
    range_lut, empty_dims, out)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   ValueError: Buffer dtype mismatch, expected 'float32_t' but got 'double'

skimage/restoration/_denoise_cy.pyx:64: ValueError
```

**Comment by hmaarrfk** [comment]
Well well, we still haven't really released 0.14.3, maybe we can skip it and go straight to 0.14.4 ;) I'll look into it this weekend.

**Comment by jni** [comment]
:sob:

**Comment by jni** [comment]
Ummmmmm...

This might be why we missed this:

https://github.com/scikit-image/scikit-image/blob/a90fd229212eddec1685318b31fc5ba915b4f4bb/skimage/restoration/tests/test_denoise.py#L187-L196

tldr we parametrize the test on image dtype but then we don't actually use the dtype parameter :man_facepalming:

**Comment by AetherUnbound** [comment]
Ah, that's almost certainly my fault D:

**Comment by jni** [comment]
@AetherUnbound many eyeballs looked over that code and thought, "looks about right!" :joy:

I think the problem is that we overly constrain the *other* arguments to be `np_floats` — we should use regular double there and cast inside the function. *I think.* Don't have time to follow up right now. =\ Anyway, I'm going to reopen. :disappointed:

**Comment by hmaarrfk** [comment]
I don't think there is fault so much as oops

**Comment by AetherUnbound** [comment]
Turns out my first forray into cython is not without Oops 😅

**Comment by bethac07** [comment]
Still getting this in 0.14.3 .  Do we expect it to be fixed, or will that be in a later release?

**Comment by hmaarrfk** [comment]
Unfortunately, I think we missed it by a day, so we need to release 0.14.4 to get the fix in. 

I think this might be a good enough reason to release 0.14.4, but I personally don't have time for it right at this second. Maybe somebody else from @scikit-image/core has the time

The PR that actually fixed things got in on May 28th
https://github.com/scikit-image/scikit-image/pull/3936

But the 0.14.3 tag was made a few days earlier.

**Comment by bethac07** [comment]
Thanks for letting me know! We're going to build/release the last CellProfiler release in the 3.X series in the next couple days, just need to figure out what skimage version to build it with.  I appreciate the explanation, makes total sense!

**Comment by hmaarrfk** [comment]
FWIW, we did discuss skipping 0.14.3 alltogether and just release 0.14.4, but it ended up being quite hard to rally enough core devs behind the idea and get it right. I was waiting to hear back of a few more issues before release 0.14.4, but this seems like it is the main one, and should get addressed.

**Comment by jni** [comment]
I've started the release process. 0.14.4 is tagged, wheels are building here:

https://travis-ci.org/scikit-image/scikit-image-wheels/builds/580668076

and here:

https://ci.appveyor.com/project/scikit-image/scikit-image-wheels
. 

Once they build I should be able to push to PyPI. But there are often issues with the wheel building, so we'll see. =\

**Comment by stefanv** [comment]
Nice work!  Linux + Mac wheels look good.  For Windows, did it build the wrong tag?  The error message looks suspicious:

https://ci.appveyor.com/project/scikit-image/scikit-image-wheels/build/job/qmurjrlss2s6jfcd

**Comment by bethac07** [comment]
I think this is the relevant build? https://ci.appveyor.com/project/scikit-image/scikit-image-wheels/builds/27172570
